### PR TITLE
#319 cache children

### LINKED ISSUE #319: Cache children
Predefine children, such as `left` and `right` in `BinaryOperator`, to get around `self.children` being slow

### PR DESCRIPTION
**Pull request by tinosulzer**
# Description

As discussed this morning, pre-assign children to get around slow anytree lookup
Fixes #319 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pybamm/expression_tree/concatenations.py [hunk 1]
```diff
@@ -30,6 +30,7 @@ def __init__(self, *children, name=None, check_domain=True):
         else:
             domain = []
         super().__init__(name, children, domain=domain)
+        self.cached_children = self.children
```
Comment (martinjrobins) [inline pybamm/expression_tree/concatenations.py]: is there a reason to only do this for binary operators and concatenations? Could simply do this for the base class `Symbol`, and then overload the `children` attribute to return `cached_children`?

Comment (tinosulzer) [reply pybamm/expression_tree/concatenations.py]: Ok, no reason, just hadn't thought of doing it that way. So if I understand correctly, we could just add a line in `Symbol.__init__` that does something like the following?
```
self.children = anytree.NodeMixin(self, children)
```

**Comment by tinosulzer** [comment]
@martinjrobins , I have moved `cached_children` to `Symbol.__init__`, but couldn't figure out how to overload the `children` property while still being able to call `NodeMixin.children` initially. If you have a good idea how to do this, could you take over the issue?

**Comment by tinosulzer** [comment]
great, thanks - good to know how to do that! Merging ...

**Pull request merged**